docs/architecture.md

```diff
@@ -1,32 +1,139 @@
+Cortex handles incoming samples using either the [distributor](#distributor) or the [query frontend](#query-frontend), depending on your Cortex setup.
 
-Finally, the query frontend will also cache query results and reuse the on subsequent queries.  If the cached results are incomplete, the query frontend will calculate the required queries and execute them in parallel on downstream queries.  The query frontend can optionally align queries with their step parameter, to improve the cacheability of the query results.
+## Services
 
-The query frontend job accepts gRPC streaming requests from the queriers, which then "pull" requests from the frontend. For HA it is recommended you run multiple frontends - the queriers will connect to (and pull requests from) all of them. To get the benefit of the fair scheduling, it is recommended you run fewer frontends than queriers - two should suffice.
+Cortex has a fundamentally service-based architecture, in which the overall system is split up into a variety of components that perform specific tasks and run separately (and potentially in parallel).
 
-See the document "[Cortex Query Woes](https://docs.google.com/document/d/1lsvSkv0tiAMPQv-V8vI2LZ8f4i9JuTRsuPI_i-XcAqY)" for more details design discussion. In the future, query splitting, query alignment and query results caching will be added to the frontend.
+### Distributor
 
-The query frontend is completely optional - you can continue to use the queriers directly.  If you want to use the query frontend, direct incoming authenticated traffic at them and set the `-querier.frontend-address` flag on the queriers.
+The **distributor** service is responsible for handling samples written by Prometheus. It's essentially the "first stop" in the write path for Prometheus samples, *unless* you run an optional [query frontend](#query-frontend), in which case the distributor takes writes from the query frontend rather than from Prometheus instances directly. Once the distributor receives samples from Prometheus, it splits them into chunks, replicates them (based on a configurable replication factor), and then sends them to the [ingester](#ingester) service.
 
-### Queriers
+Distributors communicate with ingesters via [gRPC](https://grpc.io). They are stateless and can be scaled up and down as needed.
 
-The queriers handled the actual PromQL evaluation.  They embed the chunk store client code for fetching data from long-term storage, and communicate with the ingesters for more recent data.
+#### Hashing
+
+Distributors use consistent hashing, in conjunction with the (configurable) replication factor, to determine *which* instance of the ingester service receives each sample. The hash itself is based on the metric name and tenant ID for the series. Cortex ensures that the samples for a given series all end up on the same subset of ingesters.
```

Comment:
> The hash itself is based on the metric name and tenant ID for the series

This is optional; it can be either based on metric name and tenant ID _or_ all the series labels and tenant ID.  The tradeoff with the latter is more balanced writes vs having to involve every ingester in each query.